<!DOCTYPE html>
<html>
  <head>
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.ico" />
<link href="../../css/bootstrap/bootstrap.css" rel="stylesheet" />
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link href="../../css/bootstrap/bootstrap-responsive.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="../../css/swc.css" />
<link rel="stylesheet" type="text/css" href="../../css/swc-bootstrap.css" />
<link rel="alternate" type="application/rss+xml" title="The Software Carpentry Blog" href="/feed.xml"/>
<meta http-equiv="last-modified" content="" />
<!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->

  </head>
  <body>
    <div class="container">
      <div class="banner">
  <a href="http://software-carpentry.org" title="Software Carpentry">
    <img alt="Software Carpentry banner" src="../../img/software-carpentry-banner.png" />
  </a>
</div>


      <div class="row-fluid">
        <div class="span10 offset1">
          <h1></h1>
          <h1>Title: A bit more shell</h1>

<h2>Download and open lesson material:</h2>

<p>Open your terminal or Gitbash, and enter  </p>
<div class="highlight"><pre><code class="text language-text" data-lang="text">cd                    # changes to your home directory
curl http://giladlab.uchicago.edu/data/Metahit_index.txt &gt; Metahit_index.txt
</code></pre></div>
<h2>What&#39;s going on?</h2>

<p>The curl command sends an HTTP request and dumps the result to standard out.  Here, we redirect the output of curl to <code>Metahit_index.txt</code></p>

<h2>Sanity check</h2>

<p>The above command <em>should</em> have downloaded a data table.  Let&#39;s check that the data are in the expected format--tab-delimeted text.
    head Metahit_index.txt
    4475667.3   MH0001  Denmark female  49  25.55   N   20203603    human-associated habitat
    4475668.3   MH0002  Denmark female  59  27.28   N   20203603    human-associated habitat
    ...
If this is what you see, you have the right table.  If not, you may need to troubleshoot why curl didn&#39;t give you what you were expecting.</p>

<h2>Let&#39;s move and copy the data:</h2>

<p>Let&#39;s move the data into the repo directory. This is what we would do if it were really our project: we&#39;d have a folder named &quot;Future Nature Paper&quot; and place our data in it, rather than clogging up our home directory. The command <code>mv</code> moves the data:</p>
<div class="highlight"><pre><code class="text language-text" data-lang="text">mv Metahit_index.txt ~/2014-05-12-cshl
</code></pre></div>

<p>This is clearly very important data. Let&#39;s make a back up of the data so that while we&#39;re making python scripts and running them we don&#39;t accidentally destroy the data.</p>
<div class="highlight"><pre><code class="text language-text" data-lang="text">cp Metahit_index.txt Metahit_index.bakcup.txt
</code></pre></div>
<p>While we are working through python examples in a bit, if you happen to damage your file, now you have a back up you can revert to. </p>

<h2>Let&#39;s explore the data:</h2>
<div class="highlight"><pre><code class="text language-text" data-lang="text">less Metahit_index.txt
</code></pre></div>
<p>You&#39;ll see the data has several columns: 
1. identifier 
2. another identifier 
3. country of origin
4. sex
5. age
6. BMI
7. disease status
8. PMID of the Metahit paper
9. sampling location</p>

<p>Let&#39;s look at the bottom of the file using <code>tail</code>:</p>
<div class="highlight"><pre><code class="text language-text" data-lang="text">tail Metahit_index.txt
</code></pre></div>
<p>How many individuals are in this study?</p>
<div class="highlight"><pre><code class="text language-text" data-lang="text">wc -l Metahit_index.txt
</code></pre></div>
<h2>Grep</h2>

<p>One very cool function that you can run in the shell is <code>grep</code>. <code>grep</code> is a search tool, which will search for words you enter line by line. For instance, in our data we have Danish individuals and Spanish individuals. If we want to see the data for just the Spanish individuals we can type:</p>
<div class="highlight"><pre><code class="text language-text" data-lang="text">grep Spain Metahit_index.txt
</code></pre></div>
<p>If we wanted to see only females:</p>
<div class="highlight"><pre><code class="text language-text" data-lang="text">grep female Metahit_index.txt
</code></pre></div>
<p>What happens if we want to see only males?</p>

<h2>Piping</h2>

<p>We didn&#39;t properly get to explore piping yesterday. Let&#39;s explore piping to get preliminary summary statistics on this data. Piping is essentially chaining together commands one after the other in your computer and you will get as output only the very final data coming out of the last command. So, we can use the pipe, grep, and word count command to do some interesting things:</p>

<p>How many women were in the study?</p>
<div class="highlight"><pre><code class="text language-text" data-lang="text">grep female Metahit_index.txt | wc -l
</code></pre></div>
<hr>

<p><strong>Short Exercise</strong></p>

<ol>
<li><p>Find how many Spanish people were in the study.</p></li>
<li><p>Pull out all of the information for only the first five people that did not have a disease in the dataset and save it to a file in your home directory called 5-healthy-spanish-participants.txt. </p></li>
</ol>

<hr>

<h2 id="example-text-mining-gene-expression-data">Example: text-mining gene expression data</h2>

<p>Go to data

<pre>
 cd
 cd ~/2014-05-12-cshl/materials/data/cuffdiff
</pre>
</p>

<p>This is a list of a few commands that we will use to do a little data mining of text file containing a comparison of gene expression data from and RNA-Seq experiment. </p>

<table>
  <tr><th>|</th><td><i>strings together the inputs/outputs of a series of commands</i></td></tr>
  <tr><th>cut</th><td><i>extracts sections from each line of text</i></td></tr>
  <tr><th>grep</th><td><i>searches for patterns in text</i></td></tr>
  <tr><th>sort</th><td><i>orders lines in text</i></td></tr>
  <tr><th>head</th><td><i>prints the top N lines of text</i></td></tr>
  <tr><th>wc</th><td><i>counts words, characters or lines</i></td></tr>
</table>

<h3 id="the-data">The data</h3>

<p>You have a tab-delimited text file, gene_exp.txt, that contains data from a differential gene expression analysis.  Each line describes a comparison of numerical expression levels for one gene in two samples.</p>

<h4 id="what-does-the-file-look-like">What does the file look like</h4>

<p>What is the file structure? Without options, <strong><em>head</em></strong> will print the top 10 lines of the file</p>

<pre>
$ head  gene_exp.txt
gene	sample_1	sample_2	status	value_1	value_2	significant
AT1G01010		WT		hy5	NOTEST	0	1.49367	no
AT1G01020		WT		hy5	NOTEST	7.27837	10.7195	no
AT1G01030		WT		hy5	NOTEST	1.18638	1.10483	no
AT1G01040		WT		hy5	NOTEST	0.239843	2.24208	no
AT1G01046		WT		hy5	NOTEST	0		0	no
AT1G01050		WT		hy5	OK	9.06975		23.5089	yes
AT1G01060		WT		hy5	NOTEST	4.04534		6.46964	no
AT1G01070		WT		hy5	NOTEST	1.24918		2.41377	no
AT1G01073		WT		hy5	NOTEST	0		0	no
</pre>

<p>This is a well-formatted, tab-delimited text file.  The header line describes the columns for us.  We can use this to help answer some questions.  Note that <strong><em>sort</em></strong> and <strong><em>cut</em></strong> assume that the columns in each row are tab-delimited.</p>

<h4 id="how-many-records-are-there-in-the-file">How many records are there in the file?</h4>

<p>We can use <strong><em>wc -l</em></strong> to count the lines</p>

<pre>
$ wc -l gene_exp.txt 
     200 gene_exp.txt
</pre>

<h4 id="how-many-genes-have-enough-data-to-perform-the-comparison-have-ok-status-how-many-had-significantly-different-expression-levels-between-samples">How many genes have enough data to perform the comparison (have &#39;OK&#39; status)? How many had significantly different expression levels between samples?</h4>

<table>
  <tbody>
    <tr>
      <td>We can search for <em>OK</em> and <em>yes</em> in the file, then count how many lines are returned by <strong><em>grep</em></strong>.  Note the use of the pipe symbol &#39;</td>
      <td>&#39;.  <strong><em>wc -l</em></strong> is acting on the text printed by <strong><em>grep</em></strong>, not the input file.</td>
    </tr>
  </tbody>
</table>

<pre>
$ grep OK gene_exp.txt | wc -l
    32
$ grep yes gene_exp.txt | wc -l
    12
</pre>

<p>For 199 genes, 32 had enough data to do a comparison and 12 had significantly different expression.</p>

<h3 id="question--what-if-the-strings-yes-or-ok-appear-in-other-columns-of-the-file">Question:  What if the strings <em>yes</em> or <em>OK</em> appear in other columns of the file?</h3>

<p>There is not a lot of room for free text in this example but it can&#39;t hurt to check.  This sort of thing happens all the time in real life!  We can use <strong><em>cut</em></strong> to remove the normal column (for example, column 7 for &#39;yes&#39;).  Remove column 7, then search for <em>yes</em>.</p>

<pre>
$ cut -f1-6 gene_exp.txt | grep yes
</pre>

<p>No results.  That is good.  For <em>OK</em> we need to search all columns except column 4:</p>

<pre>
cut -f1-3,5-7 gene_exp.txt | grep OK
</pre>

<p>No results.  The file is good.  Note that the <strong><em>-k</em></strong> argument could also have been expressed as <strong><em>-k1,2,3,5,6,7</em></strong></p>

<h4 id="what-are-the-20-genes-with-the-highest-expression-levels-in-sample-1-and-differ-significantly-between-samples">What are the 20 genes with the highest expression levels in sample 1 and differ significantly between samples?</h4>

<p>We can use <strong><em>grep</em></strong> to get the &#39;yes&#39; lines, then use <strong><em>sort</em></strong> to order the lines base on the numeric values in column 5 (value_1).  We pipe the output to <strong><em>head</em></strong> so we just look at the top 10 lines for now.  The <strong><em>k5</em></strong> argument means sort on column (key) 5; <strong><em>-n</em></strong> means sort numerically.  </p>

<pre>
$ grep 'yes' gene_exp.txt | sort -k5 -n | head
AT1G01320	WT	hy5	OK	4.94764	20.8172	yes
AT1G01050	WT	hy5	OK	9.06975	23.5089	yes
AT1G02120	WT	hy5	OK	11.3678	23.8411	yes
AT1G01610	WT	hy5	OK	14.4058	5.2457	yes
AT1G01120	WT	hy5	OK	15.6414	7.70519	yes
AT1G01430	WT	hy5	OK	19.7868	10.301	yes
AT1G01090	WT	hy5	OK	55.3618	33.6315	yes
AT1G01170	WT	hy5	OK	71.5683	21.235	yes
AT1G02500	WT	hy5	OK	71.7299	32.6785	yes
AT1G02140	WT	hy5	OK	105.425	65.1563	yes
</pre>

<p>You may have noticed that cut and sort use different arguments for the same thing (column number).  The collection of tools in unix-like operating systems evolved over time from a variety of sources and authors, so their command line arguments are not always consistent.  If in doubt:</p>

<pre>
man cut
</pre>

<p>Note the the values in column 5 above are all zeros.  We are not quite there yet.  We can use the <strong><em>r</em></strong> flag to sort in descending order.</p>

<pre>
$ grep 'yes' gene_exp.txt | sort -k5 -n -r | head
AT1G01100	WT	hy5	OK	275.52	192.323	yes
AT1G01620	WT	hy5	OK	178.441	80.266	yes
AT1G02140	WT	hy5	OK	105.425	65.1563	yes
AT1G02500	WT	hy5	OK	71.7299	32.6785	yes
AT1G01170	WT	hy5	OK	71.5683	21.235	yes
AT1G01090	WT	hy5	OK	55.3618	33.6315	yes
AT1G01430	WT	hy5	OK	19.7868	10.301	yes
AT1G01120	WT	hy5	OK	15.6414	7.70519	yes
AT1G01610	WT	hy5	OK	14.4058	5.2457	yes
AT1G02120	WT	hy5	OK	11.3678	23.8411	yes
</pre>

<p>OK, now we have the 10 most abundant genes in sample 1.  Since we were only asked to list the genes, we can use <strong><em>cut</em></strong> to get just the list of gene names</p>

<pre>
$ grep 'yes' gene_exp.txt | sort -k5 -n -r | head | cut -f1
AT1G01100
AT1G01620
AT1G02140
AT1G02500
AT1G01170
AT1G01090
AT1G01430
AT1G01120
AT1G01610
AT1G02120
</pre>

<p>And we have our answer!</p>


	</div>
      </div>

      <div class="footer">
  <a class="label swc-blue-bg" href="https://www.facebook.com/SoftwareCarpentry">Facebook</a>
  <a class="label swc-blue-bg" href="https://plus.google.com/u/0/114244759874490019250/posts">Google+</a>
  <a class="label swc-blue-bg" href="https://twitter.com/swcarpentry">Twitter</a>
  <a class="label swc-blue-bg" href="https://github.com/swcarpentry">GitHub</a>
  <a class="label swc-blue-bg" href="http://software-carpentry.org/feed.xml">RSS</a>
  <a class="label swc-blue-bg" href="../../license.html">License</a>
  <a class="bugreport label swc-blue-bg" href="mailto:info@software-carpentry.org?subject=bug%20in%20lessons/thw-shell/moreshell.md">Bug Report</a>
</div>

    </div>
    <!-- Le javascript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="../../js/jquery-1.9.1.min.js"></script>
<script src="../../js/bootstrap/bootstrap.min.js"></script>

  </body>
</html>
